### ISSUE TITLE
Web of Things server test framework?

### ISSUE THREAD

**Issue by draggett** [w3c/web-of-things-framework#16]
We need common test framework to verify interoperability of servers by ensuring that they pass a common test suite. 

The tests will need to cover at least:

* The basic model of things in terms of metadata, events, properties and actions
* Dependencies between things hosted on the same or different servers
* Security and privacy related features, e.g. authentication and access control
* Support to multiple protocols, e.g. HTTP, WebSockets, CoAP, MQTT, XMPP and AMQP.

We will need a declarative format for tests, and a mechanism to adding, documenting and noting the status of tests, e.g. reviewed, what bugs a test has, and so forth. Likewise, it would be very desirable to have a registry of test results across servers. 

One suggestion is that it may be worth considering writing a test harness in NodeJS given the availability of node modules for many protocols.  Please respond to this issue if you would like to contribute to work on the test framework in any way.

**Comment by draggett** [comment]
Joachim suggest looking at Jenkins and Travis-Cl, see:
* https://stackoverflow.com/questions/14274293/show-current-state-of-jenkins-build-on-github-repo
* https://wiki.jenkins-ci.org/display/JENKINS/Git+Plugin
* https://stackoverflow.com/questions/13546097/how-do-show-my-tests-passing-failing-in-github

Anyone care to study these and provide a summary of their applicability?

**Comment by mcicolella** [comment]
Hi, I have some interesting links
- TeamCity CI http://ariya.ofilabs.com/2015/03/continuous-integration-for-node-js-projects-with-teamcity.html
- Travis CI & Heroku http://shapeshed.com/continuously-deploy-node-apps-with-github-travis-and-heroku/

The last is a simple and complete guide to set an account on Travis CI (free for open source public project) and deploy on Heroku.
This is their proposed build and deployment pipeline
1. Hack, hack, hack
2. Git Commit Locally
3. Git Push to GitHub
4. A build is triggered on Travis
5. If tests fail nothing happens
6. If tests pass a deployment to heroku is triggered
7. Your app is deployed to heroku by travis

**Comment by mpetyx** [comment]
I really like using Travis in various open source projects.
It is indeed a pretty cool tool.
I would further recommend adding a Behaviour Driven Development (BDD) tool like [cucumber](https://github.com/cucumber/cucumber-js). The most important aspect in this framework is the [Gherkin language](https://github.com/cucumber/cucumber/wiki/Gherkin) which makes it easy to create tests out of the user stories.
This can also be combined with issue #3 .
The workflow as I imagine it could be:
* Describe a new User Story.
* Define the corresponding Gherkin File.
* Create the equivalent unittest. 
* Validate that it works also with the interface defined in #3 .
* Pull/Push the changes.
* Then have the workflow defined in https://github.com/w3c/web-of-things-framework/issues/16#issuecomment-109881729 triggered.

**Comment by bubbafat** [comment]
Is this issue about testing the reference node implementation or providing a test suite that can execute against arbitrary servers using only the HTTP/WS interfaces?

I get the impression that what dsr is asking for is a tool to validate servers written by other parties - if I'm misunderstanding that then ignore the rest of this.

I'm not entirely sure I'm understanding how CI tools are going to help with the issue.

If I'm going to write my own WoT server implementation I'm not going to care about what processes or tools are being used here - I'm going to use my own.  Along the way, and especially at the end, I want to have a test suite I can run that uses only the HTTP and WS interfaces to ensure that I'm behaving in expected ways.

I don't care about Travis or Heroku or any of the other (great) systems.  I just want to execute

./validate-wot http://localhost:8888/

Or just visit http://wotvalidator.org/ and enter a url for my service.

Obviously a command line version will need to have some dependency - I'd prefer something like python that exist pretty much everywhere - though a node module would probably be faster to get going initially.

Or am I totally missing the point?

**Comment by nweedon** [comment]
I agree with @bubbafat that the question seems more aimed towards testing against arbitrary servers to determine their levels of access / conformance to the standard (across different protocols, security, data format etc.), rather than testing a reference implementation with CI tools.

If this is the case, building from the comment above, the validation process could be determined by framing the problem in the context of a device or 'thing' connecting to the service (albeit one that supports all protocols/formats etc.), in which case, using Mocha, Jasmine or another testing framework could be a beneficial step if the validation is carried out with Node.js. However, if this is the direction to be taken, one of the first steps would be to make sure that there are node packages available for each of the protocols the standard will support.

**Comment by draggett** [comment]
That is a goal for the NodeJS project. In the short term, we can create test scripts that can be run with the protocol bindings that exist and exercise the error handling.  I see this as a evolutionary process that will drive improvements to the code, and in turn suggest additional test cases.
